File: native/app/lib/capture_scheduler.dart
import 'dart:async';

/// 撮影の周期実行。`Timer` をここに閉じ込め、多重起動の防止と停止を保証する。
///
/// [CaptureSession] がタイマーを持たない代わりに、周期実行の責務はこのクラスが
/// 単独で担う。`Timer` を直接扱う場所を1か所に限定することで、キャンセル漏れ
/// （停止したのに撮影が続く）を起こしにくくする。
class CaptureScheduler {
  Timer? _timer;

  bool get isActive => _timer != null;

  /// [interval] ごとに [onTick] を呼ぶ。すでに動作中なら何もせず false を返す。
  ///
  /// 開始直後にも1回発火させる（`AGENTS.md` 5.2-4 の「開始でタイマー起動」を、
  /// 最初の1枚を待たせずに撮る挙動として実装している）。
  bool start(Duration interval, void Function() onTick) {
    if (_timer != null) return false;
    _timer = Timer.periodic(interval, (_) => onTick());
    onTick();
    return true;
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  void dispose() => stop();
}

File: native/android/app/src/main/kotlin/com/kotonara/farmcamera/domain/CaptureScheduler.kt
package com.kotonara.farmcamera.domain

import kotlin.time.Duration

/**
 * 一定間隔でコールバックを発火させる抽象（docs/03-native.md 7 節）。
 *
 * 実装は `data` 層に置く。ここを interface にしておくことで、撮影ループを回す側を
 * 実機なしでテストできる。
 */
interface CaptureScheduler {
    /** 動作中か。停止後・開始前は false。 */
    val isActive: Boolean

    /**
     * [interval] ごとに [onTick] を発火させる。**開始した時点で即 1 回発火する。**
     *
     * 次のサイクルまで待たせると、デモで「開始したのに数分間何も起きない」状態になる。
     *
     * 開始できなければ `false` を返す。動作中の再呼び出しを拒否するのは、
     * タイマーが多重起動して撮影が二重に走るのを防ぐため。
     * ゼロ以下の [interval] も拒否する。受け付けるとタイマーが暴走する。
     */
    fun start(
        interval: Duration,
        onTick: () -> Unit,
    ): Boolean

    /** 停止する。停止後は発火しない。 */
    fun stop()
}
